File: src/test/java/it/xpug/kata/birthday_greetings/BirthdayServiceTest.java
package it.xpug.kata.birthday_greetings;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.text.ParseException;
import java.util.Collections;

import static java.util.Collections.singletonList;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class BirthdayServiceTest {

    public static final String BIRTHDAY_DATE = "2008/10/08";
    public static final String ANOTHER_DATE = "2008/01/01";

    @Mock
    private Sender sender;

    @Mock
    private EmployeesRepository employeesRepository;

    @Test
    public void should_send_message() throws Exception {
        BirthdayService birthdayService = new BirthdayService(employeesRepository, sender);

        when(employeesRepository.all()).thenReturn(singletonList(anEmployWith(BIRTHDAY_DATE)));

        birthdayService.sendGreetings(new XDate(BIRTHDAY_DATE));

        verify(sender).sendGreetingsTo(any(Employee.class));
    }

    @Test
    public void should_not_send_message() throws Exception {
        BirthdayService birthdayService = new BirthdayService(employeesRepository, sender);

        when(employeesRepository.all()).thenReturn(singletonList(anEmployWith(BIRTHDAY_DATE)));

        birthdayService.sendGreetings(new XDate(ANOTHER_DATE));

        verify(sender, never()).sendGreetingsTo(any(Employee.class));
    }

    @Test
    public void should_not_send_message_without_employees() throws Exception {
        BirthdayService birthdayService = new BirthdayService(employeesRepository, sender);
        when(employeesRepository.all()).thenReturn(Collections.<Employee>emptyList());

        birthdayService.sendGreetings(new XDate("2008/10/08"));

        verify(sender, never()).sendGreetingsTo(any(Employee.class));
    }

    private Employee anEmployWith(String birthdayDate) throws ParseException {
        return new Employee(null, null, birthdayDate, null);
    }
}

File: src/test/java/it/xpug/kata/birthday_greetings/BirthdayServiceTest.kt
package it.xpug.kata.birthday_greetings

import com.nhaarman.mockito_kotlin.*
import org.junit.Test

class BirthdayServiceTest {
    private val NOT_TODAY_STRING = "2018/08/08"
    private val TODAY_STRING = "2019/09/09"
    private val TODAY = XDate(TODAY_STRING)
    private val sender : Sender = mock()
    private val repository : EmployeeRepository = mock()

    @Test
    fun `without employees do not send greetings`() {
        whenever(repository.all()).thenReturn(emptyList())

        BirthdayService(sender, repository).sendGreetings(TODAY)

        verify(sender, never()).send(any(), any(), any(), any())
    }

    @Test
    fun `without employees born today do not send greetings`() {
        whenever(repository.all()).thenReturn(listOf(bornAnotherDayEmployee("another@day.local")))

        BirthdayService(sender, repository).sendGreetings(TODAY)

        verify(sender, never()).send(any(), any(), any(), any())
    }

    @Test
    fun `with one employee born today send greetings once`() {
        whenever(repository.all()).thenReturn(listOf(bornTodayEmployee("Marco", "today@day.local")))

        BirthdayService(sender, repository).sendGreetings(TODAY)

        verify(sender).send(
            "sender@here.com",
            "Happy Birthday!",
            "Happy Birthday, dear Marco",
            "today@day.local"
        )
    }

    @Test
    fun `with two employee, one born today send greetings once`() {
        whenever(repository.all()).thenReturn(listOf(
            bornAnotherDayEmployee("another@day.local"),
            bornTodayEmployee("Marco", "today@day.local")
        ))

        BirthdayService(sender, repository).sendGreetings(TODAY)

        verify(sender).send(any(), any(), any(), eq("today@day.local"))
    }

    @Test
    fun `with three employee, two born today send greetings twice`() {
        whenever(repository.all()).thenReturn(listOf(
            bornAnotherDayEmployee("a-normal@day.local"),
            bornTodayEmployee("Marco", "today1@day.local"),
            bornTodayEmployee("Sonia", "today2@day.local")
        ))

        BirthdayService(sender, repository).sendGreetings(TODAY)

        verify(sender).send(any(), any(), any(), eq("today1@day.local"))
        verify(sender).send(any(), any(), any(), eq("today2@day.local"))
    }

    private fun bornAnotherDayEmployee(email: String) = Employee("name", "surname", NOT_TODAY_STRING, email)

    private fun bornTodayEmployee(name: String, email: String) = Employee(name, "surname", TODAY_STRING, email)
}
